cpp/src/codegen/types.cc

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2017-2018 Dremio Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gandiva/types.h"
+
+#include "codegen/function_registry.h"
+#include "codegen/llvm_types.h"
+
+namespace gandiva {
+
+DataTypeVector Types::supported_types_ = Types::InitSupportedTypes();
+
+const FuncSignatureVector Types::supported_functions() {
+  return FunctionRegistry::GetSupportedFunctions();
+}
+
+DataTypeVector Types::InitSupportedTypes() {
+  DataTypeVector data_type_vector;
+  llvm::LLVMContext llvm_context;
+  LLVMTypes llvm_types(llvm_context);
+  auto supported_arrow_types = llvm_types.GetSupportedArrowTypes();
+  for (auto &type_id : supported_arrow_types) {
+    AddArrowTypeToVector(type_id, data_type_vector);
+  }
+  return data_type_vector;
+}
+
+void Types::AddArrowTypeToVector(arrow::Type::type &type, DataTypeVector &vector) {
+  switch (type) {
```

Comment:
can you use type_traits ?

auto type = TypeTraits\<TypeParam\>::type_singleton();

Reply:
as discussed that cannot be used since it needs a data type as well..